Add support for distributed functions in extensions of distributed actors

DougGregor opened the pull request:
Fix a few minor issues in the type checker and SILGen to properly cope with
distributed functions defined within extensions of distributed actors.
While here, centralize the logic that adds the "_remote_" function.

... and since this uncovered a problem with the mangling for distributed
functions, fix that, too.

Fixes rdar://84325525.

DougGregor committed:
Add support for distributed functions in extensions of distributed actors.

Fix a few minor issues in the type checker and SILGen to properly cope with
distributed functions defined within extensions of distributed actors.
While here, centralize the logic that adds the "_remote_" function.

Fixes rdar://84325525.

DougGregor committed:
Add test case

DougGregor commented:
@swift-ci please smoke test and merge

ktoso commented on lib/Sema/CodeSynthesisDistributedActor.cpp:
```diff
@@ -189,7 +188,24 @@ AbstractFunctionDecl *TypeChecker::addImplicitDistributedActorRemoteFunction(
+    return nullptr;
+
+  return addImplicitDistributedActorRemoteFunction(DC, func);
+}
```
thanks for the cleanup here

ktoso reviewed (approved):
Looks good, thanks a lot!

DougGregor committed:
Use a non-conflicting mangling for distributed thunks.

Distributed thunks were using the same mangling as direct method
reference thunks (i.e., for "super" calls). Although not technically
conflicting so long as actors never gain inheritance, it's confusing
and could cause problems in the future. So, introduce a distinct
mangling for distributed thunks and plumb them through the demangling
and remangler.

ktoso commented on docs/ABI/Mangling.rst:
```diff
@@ -217,6 +217,7 @@ types where the metadata itself has unknown layout.)
   global ::= global 'To'                 // swift-as-ObjC thunk
   global ::= global 'TD'                 // dynamic dispatch thunk
   global ::= global 'Td'                 // direct method reference thunk
+  global ::= global 'TE'                 // distributed actor thunk
```
okey E then! Thanks for spotting this

ktoso replied:
Hmmm we also used the `Td` in:

```

  case Kind::DistributedThunkAsyncFunctionPointer: {
    std::string Result = getSILDeclRef().mangle();
    Result.append("Td");
    Result.append("Tu");
    return Result;
  }

```

which this doesnt fix yet -- i think there we also need to use TE right? I'll follow up with this.

ktoso commented:
@swift-ci please build toolchain macOS

ktoso commented:
Actual failure, I think I know how to solve -- on it.

ktoso commented:
Replaced by https://github.com/apple/swift/pull/39787 where the only additional fix is https://github.com/apple/swift/pull/39787/commits/e1ae0be2dd1b45fd5aec75a033338279e2b89b83
